### ISSUE TITLE
Kotlin nullness analysis integration

### ISSUE THREAD

**Issue by wmdietlGC** [jspecify/jspecify#24]
Information at:
- https://github.com/JetBrains/kotlin/blob/master/spec-docs/flexible-java-types.md
- https://kotlinlang.org/docs/reference/java-interop.html#null-safety-and-platform-types
- A talk explaining platform types in Kotlin + motivational cases: [Video](https://www.youtube.com/watch?v=2IhT8HACc2E), [Slides](https://www.dropbox.com/s/p5cqn74q7lqfpl3/Flexible%20Types%20in%20Kotlin%20-%20JVMLS%202015.pdf?dl=0)

**Comment by PetukhovVictor** [comment]
I see the following way for integration nullness annotations into Kotlin. Keep in mind we should avoid breaking changes between minor versions anyway and major versions without preliminary warnings.

So the integration plan can be like this:
1) (as soon as possible) Start reporting warnings in the Kotlin use sites affected by the new jspecify annotations about future possible change of resolution or appearing a red code. We can do efforts to understand where the resolution can really change or applicability can be broken and report the warnings only there.
2) (as soon as possible, partially done) Check the current implementation of support jspecify annotations by @dzharkov and merge it into master under an internal or public flag.
3) (as soon as possible, partially done) Fix an issue around support loading type use annotations from Java binaries (the corresponding issue on the IntelliJ side has already been resolved): https://youtrack.jetbrains.com/issue/KT-11454
4) (anytime up to 1.5) Fix a few corner cases' problems and implement some design decisions, which are related with support of jspecify annotations (see our internal document "Type annotations support in Kotlin").
5) (in 1.5) Enable jspecify annotations support by default.

Also taking into account this:
> One topic that we've discussed occasionally is what happens to Kotlin users of a library like Guava when all of Guava's jsr305 and Checker Framework annotations are replaced with our new annotations. One thing that can happen is that they see new compile errors (or warnings, if the new checks in Kotlin are limited to warning-level). Here's an example of how that can happen.

I think we can at least detect the situation when used version of the Kotlin compiler is less than 1.5 and the new annotations are used in the version of the library known by us. In such case, the resolution can really change and new errors can appear: the library **no longer uses** the old annotations, and the Kotlin compiler **doesn't yet support** the new ones by default. For instance, we can report this using some IDE-level warning and suggest turn on the new annotations interpretation through a compiler flag or update the Kotlin compiler/downgrade a library. Otherwise, warnings only on call sites may be missed.

So it seems there is another plan item:
6) (at the same time or earlier with the start of the warnings reporting) Implement additional detection and suggestion mechanism for the case above.

@abreslav What do you think about such plan? Maybe I missed something or wrote something wrong.

Also @cpovirk, thanks for described cases, do you think this fits into the plan above?

**Comment by cpovirk** [comment]
Since I keep letting this sit: @kevin1e100 , you probably have a better idea of what we need for Kotlin than I do. Would you mind having a look?

**Comment by kevin1e100** [comment]
Sounds very reasonable to me.

> we can report this using some IDE-level warning and suggest turn on the new annotations interpretation through a compiler flag or update the Kotlin compiler/downgrade a library.

This sounds fine to me, but if JSpecify annotations will be errors starting in 1.5 then I suspect there's a good chance we won't have to worry about this case. That's just my intuition but I have no objections to implementing some safeguards around this regardless.

**Comment by PetukhovVictor** [comment]
Some update.

Unfortunately, in fact, in Kotlin we will not be able to enable error reporting for jspecify annotations by default in 1.5, as this does not comply with our breaking changes policy, because of jspecify annotations was supported only in 1.4.30 (in the warn mode by default), until 1.5 it will take too little time to enable the error level by default. I am also currently finishing work on support for reading type use annotations from class files and some other improvements (for example, correctly reading annotations on arrays: `@Anno1 Integer @Anno2 [] @Anno3 []`), which is also a breaking change. These commits will only get into 1.5. At the same time, proper support for reading type use annotations is important for jspecify cases, so I assumed it would be nice if these features (improvements plus jspecify annotation support itself) had the same life cycle so that support was complete and consistent at every point in time. So I see the following scenario: in 1.5 it will be supported to read jspecify annotations properly from both sources and binaries in the warn mode by default with the ability to enable the error mode via a compiler flag; and in 1.6 enable the error mode by default.

It's also possible that the minimum time for adoption (time to enable errors by default) will be expressed not in versions ("from 1.5 to 1.6"), but in time (for example, half a year). I will additionally inform you if something changes here.

**Comment by cpovirk** [comment]
@PetukhovVictor , a question for you (or maybe your IntelliJ colleagues):

Once your work on the Kotlin compiler is done, will IntelliJ's _Kotlin_ editor support the same JSpecify warnings/errors as the command-line compiler does? Or does that require additional work?

(I understand the IntelliJ's _Java_ editor would need significant work to catch up to the Kotlin side.)

**Comment by PetukhovVictor** [comment]
> Once your work on the Kotlin compiler is done, will IntelliJ's Kotlin editor support the same JSpecify warnings/errors as the command-line compiler does?

Yes, it will, but not fully in all combinations of IntelliJ's version and Kotlin plugin's version:
- IJ 20.1/20.2 + Kotlin 1.4.30: it will work only with sources and without some improvements of the type enhancement (like [KT-24392](https://youtrack.jetbrains.com/issue/KT-24392), [KT-43919](https://youtrack.jetbrains.com/issue/KT-43919), [KT-43920](https://youtrack.jetbrains.com/issue/KT-43920) and some more);
- IJ 20.3 + Kotlin 1.4.30: it will work both with sources and binaries in the editor and only with sources during compilation (in a sense, these two versions are not fully compatible in terms of loading Java annotations), type enhancement improvements aren't available;
- IJ 20.3 + Kotlin 1.5.0: it will work both with sources and binaries fully, with type enhancement improvements;
- IJ 20.1/20.2 + Kotlin 1.5.0: it will work both with sources and binaries during compilation, and only with sources in the editor, type enhancement improvements are available (except for support related reading type use annotations from binaries).

The differences are due to the fact that compilation and the IntelliJ editor use different Java models (note that in these models at different times there was support for reading type use annotations from binaries).

Additional work on the IntelliJ side to support something around jspecify annotations in the Kotlin's editor aren't needed for now. [The main task](https://youtrack.jetbrains.com/issue/IDEA-153093), important for us, has already been done and are available in IJ 20.3.

**Comment by cpovirk** [comment]
Thanks! I had figured, but I wanted to make sure I wasn't overlooking something.

**Comment by PetukhovVictor** [comment]
As we discussed internally earlier, support for type use annotations and some other improvements will be available in 1.4.30 (the corresponding type mismatch reports will be warnings, the compiler flag `-XXLanguage:+ImprovementsAroundTypeEnhancement` turns them to errors), so in the post above it's worth replacing `1.5.0` with `1.4.30` and ignoring the points where `1.4.30` was before :)

BTW, "some other improvements" are:
*  use annotations in the type parameter position to enhance corresponding types: [KT-11454](https://youtrack.jetbrains.com/issue/KT-11454)
*  proper support of the type enhancement of the annotated java arrays: [KT-24392](https://youtrack.jetbrains.com/issue/KT-24392)
*  proper support of the type enhancement of the annotated java varargs' elements: [KT-18768](https://youtrack.jetbrains.com/issue/KT-18768)
*  type enhancement based on annotated bounds of type parameters
*  type enhancement within type arguments of the base classes and interfaces: [KT-43919](https://youtrack.jetbrains.com/issue/KT-43919)
*  support type enhancement based on type use annotations on java fields
*  preference of a type use annotation to annotation of another type: [KT-24392](https://youtrack.jetbrains.com/issue/KT-24392)
    * (if @NotNull has TYPE_USE and METHOD target, then `@NotNull Integer []` -> `Array<Int>..Array<out Int>?` instead of `Array<Int>..Array<out Int>`)

**Comment by cpovirk** [comment]
Thanks! I had just yesterday built a kotlinc from the 1.4.30 branch and started experimenting with `-XXLanguage:+ImprovementsAroundTypeEnhancement -Xjspecify-annotations=strict`, based on your earlier instructions.

Based on my experiments, I had hoped to send a question to you today, but it looks like I might not get to it until the new year.

**Comment by cpovirk** [comment]
(no rush on this; I'm just trying to dump some thoughts from a recent conversation before I forget them)

One question that's come up in recent discussions is whether it's desirable to have Kotlin issue only warnings (i.e., not errors) for a while _after_ we release JSpecify 1.0 and projects begin to adopt it.

A small reason for this approach is that we'd get further testing of Kotlin's support before it's on "for real."

But my impression is that the bigger reason is that library users would get a transition period for free: If we don't provide library owners with [a `@Migrating` annotation](https://github.com/jspecify/jspecify/issues/26) in the 1.0 release, then by default, either all JSpecify annotations can produce errors, or all JSpecify annotations can produce warnings. Since it can be useful to offer a transition period in which problems are reported only as warnings, Kotlin could choose to provide that for all JSpecify annotations in the early days of 1.0.

(Yes, users can always switch between errors and warnings by changing compiler flags (or doing whatever the equivalent of that is in their IDE, I assume?). But the default behavior matters a lot.)

(And yes, we'd still hope to eventually provide features to simplify migration long after the release of 1.0, since that's when most people would adopt. But such features may take time to design, so we may conclude that it's best to release without them, in which case we'd want to consider "only warnings" as a short-term solution for Kotlin users of early-adopter libraries.)

I see value in this, but I want to note one reason that I don't see it as a slam dunk: Yes, it provides a valuable transition period when someone annotates a previously unannotated library -- or at least adds more nullness information to a library that starts out with partial nullness information. But I see a disadvantage for libraries that are already annotated and are looking to switch to JSpecify annotations: If JSpecify annotations produce only warnings, then a switch to JSpecify is likely to downgrade nullness problems from errors to warnings. Given that, I would expect some libraries not to want to do that.

That said, libraries can choose to use multiple sets of annotations during a transition period -- and that might be a good idea for other reasons. ([Guava will probably do this.](https://github.com/jspecify/jspecify/issues/239)) So a library could stick with an older set of annotations (one that produces errors by default) while also adding the JSpecify annotations (even if they produce only warnings) to signal that they're shifting in that direction. Maybe we'll even have `@Implies` by then so that a single annotation can serve both roles.

**Comment by cpovirk** [comment]
Some notes from a chat with kevinb9n:
- As of the recent [1.8.20](https://github.com/JetBrains/kotlin/releases/tag/v1.8.20) release, Kotlin supports `org.jspecify.annotations` to replace `org.jspecify.nullness`.
- Kotlin's JSpecify support has only one bug that worries me much, [KT-50008](https://youtrack.jetbrains.com/issue/KT-50008). I remember also encountering [KT-55388](https://youtrack.jetbrains.com/issue/KT-55388) along the way, though that's not JSpecify-specific. Finally, it would also be good to eventually deal with [KT-53836](https://youtrack.jetbrains.com/issue/KT-53836).
- Kotlin will still need support for `@NonNull` ([KT-53835](https://youtrack.jetbrains.com/issue/KT-53835)) and `@NullUnmarked` ([KT-53834](https://youtrack.jetbrains.com/issue/KT-53834)).
- We'd also like for Kotlin to _emit_ JSpecify annotations ([KT-47417](https://youtrack.jetbrains.com/issue/KT-47417)), though we may be able to find [alternative ways to get that information](https://docs.google.com/document/d/1ZBGwQ4I81-Jx7HDqYlW77aRd-pmwlW_oNkUCRagx2yE/edit) for at least some of our nullness checking.

The support continues to work well for us inside Google. We've seen some of the concerns from the list above crop up occasionally, and we'd expect that to increase over time, but so far, so good.

**Comment by cpovirk** [comment]
I jinxed it! :)

I just filed [KT-57996](https://youtrack.jetbrains.com/issue/KT-57996) for a problem with annotations on arrays. As I note in that bug, I suspect that the Kotlin people are already halfway to fixing it.

**Comment by kevinb9n** [comment]
@mglukhikh Could you give us an overall update on where things stand with Kotlin and JSpecify compliance?

I'll also note that making our conformance test suite (ready soon) be useful to Kotlin will take some extra time and may need some extra help. We will do our best to separate sample APIs from sample clients so that they can be ported separately and the cross-traffic tested.

**Comment by cpovirk** [comment]
We should also figure out what the status of `-Xtype-enhancement-improvements-strict-mode` is and how much that status matters to JSpecify. I just said [a little more about that in another thread](https://github.com/jspecify/jspecify/issues/239#issuecomment-1545817186).

**Comment by cpovirk** [comment]
Here's a bug we just encountered for the first time. It's not specific to JSpecify, but it's coming up for us as part of annotating Java classes with JSpecify annotations: [KT-61227](https://youtrack.jetbrains.com/issue/KT-61227).

**Comment by z0mb1ek** [comment]
hi! [Caffeine library](https://github.com/ben-manes/caffeine) version 3.2.0 started using the @Nullable annotation and kotlin version 2.1.21 stopped compiling. Is there any way I can fix this?
